Support complete WFTDA statsbook

brian-brazil opened the pull request:
To make review easier...

frank-weinberg committed:
Turn Event properties into enums

frank-weinberg committed:
streamline XML element retrieval

also fix a bug in the PT screen, where all skaters where shown as fielded

frank-weinberg committed:
add enum values for values not sent via events

frank-weinberg committed:
unify (most of) JSON path generation

frank-weinberg committed:
Make removal explicit attribute of relevant Events

frank-weinberg committed:
mostly unify handling of permanent properties

penalties and rulesets need special treatment

frank-weinberg committed:
align penalty handling with other elements

frank-weinberg committed:
Align Rulesets and penalty codes with other elmnts

Also fix some bugs

frank-weinberg committed:
unify handling of CommandProperties

frank-weinberg committed:
unify handling of AddRemoveProperties

simplify generating and parsing XML and JSON
align naming of MediaFormat and MediaType in XML/JSON with other types
print out exceptions from XML parsing
fix some bugs uncovered by exception printout

frank-weinberg committed:
add support to set any SB element through JSON

migrate path for Settings in JSON to other element's syntax

frank-weinberg committed:
bugfixes & code cleanup

frank-weinberg committed:
migrate ruleset editor to generic WS set interface

frank-weinberg committed:
migrate PT screen to generic WS commands

frank-weinberg committed:
add explanatory comments to the functions of ScoreBoardEventProvider

frank-weinberg committed:
move Stats elements into separate classes

rename PeriodStats to Period
rename JamStats to Jam
rename TeamStats to TeamJam
rename SkaterStats to Fielding

adjust frontend names to new names where they don't match already

frank-weinberg committed:
integrate Period

- make periods direct children of ScoreBoard
- store inPeriod as currentPeriod.running
- add begin and end time to period

frank-weinberg committed:
add infrastructure for numbers as id

use it for periods
also switch usage of getId() and getProviderId()

frank-weinberg committed:
integrate Jam

also move some more generic functionality into ScoreBoardEventProvider

frank-weinberg committed:
make Clock, Position, and Team Tests use full ScoreBoard

frank-weinberg committed:
keep current Period and Jam number in the respective parent classes

frank-weinberg committed:
Integrate TeamJam and Fielding into rest of core

frank-weinberg committed:
make penalty class standalone and integrate it with other elements

frank-weinberg committed:
Add interfaces to add/remove periods and jams.
Fix eclipse indentaiton setting and indentations.
Further extend sbep framework,

frank-weinberg committed:
add support for full SK stats

frank-weinberg committed:
add support for box trips

frank-weinberg committed:
quick on-site-fixes from first game

frank-weinberg committed:
Turn references into ScoreBoardListeners.
Store data type and default value with properties. Enforce type.

frank-weinberg committed:
add SK-sheet displays to operator screen to allow modification of scring
trips

frank-weinberg committed:
add LT sheet screen

frank-weinberg committed:
add PLT screen (untested)

frank-weinberg committed:
complete plt screen

frank-weinberg committed:
quick post scrimmage fixes

frank-weinberg committed:
turn new lt, sk, and plt from standalone pages into modules. Add page
for PLT screen.

frank-weinberg committed:
fix clearing of Fields in XML.

frank-weinberg committed:
various small fixes and improvements

frank-weinberg committed:
fix deleting jams and box trips

frank-weinberg committed:
remove NPE when moving the end of a box trip into the previous jam

brian-brazil commented on config/penalties/wftda2018.json:
```diff
@@ -1,5 +1,6 @@
 {
 	"penalties": [
+		{ "code": "?", "verbalCues": ["Unknown"]},
```
What about the other file?

brian-brazil commented on html/controls/operator.js:
```diff
@@ -765,22 +764,62 @@ function createTeamTable() {
+			niButton.toggleClass("Active", isTrue(value));
+		});
+		niButton.appendTo(leadJammerTd);
 		/* some strange bug, css direction is unset for leadJammerTd
```
Is this still the case?

brian-brazil commented on html/controls/operator.js:
```diff
@@ -892,6 +928,163 @@ function createTimeTable() {
+		}
+	});
+	
+	WS.Register(['ScoreBoard.CurrentPeriodNumber'], function(k, v) {
```
This is a bit unsafe, if we want exact matches we should add a function for that.

frank-weinberg replied:
That sounds worthy of a separate PR.

brian-brazil replied:
Sure

brian-brazil commented on html/controls/plt/plt.js:
```diff
@@ -0,0 +1,477 @@
+	function penaltyCode(k, v) {
+		var match = (k || "").match(codeIdRegex);
+		if (match == null || match.length == 0)
+			return;
```
Please use explicit braces here

brian-brazil commented on html/json/WS.js:
```diff
@@ -93,11 +93,12 @@ var WS = {
 	},
 
-	Set: function(key, value) {
+	Set: function(key, value, flag = '') {
```
Hmm, what browsers do we support? Default values aren't that old.

frank-weinberg replied:
Officially we supprt Firefox and Chrome. I don't think we give a minimum version anywhere.

brian-brazil replied:
Default values only started being a thing in 2015. I don't think it's safe to assume everyone has that.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/core/BoxTrip.java:
```diff
@@ -0,0 +1,62 @@
+        START_AFTER_S_P(Boolean.class, false),
+        END_FIELDING(Fielding.class, null),
+        END_JAM_NUMBER(Integer.class, 0),
+        END_BETWEEN_JAMS(Boolean.class, false),
```
What's this for?

frank-weinberg replied:
Box Trips ending between Jams, e.g. because the Penalty is removed.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/core/Clock.java:
```diff
@@ -20,12 +22,10 @@
 
     public String getName();
-    public void setName(String name);
+    //    public void setName(String name);
```
Can it be removed?

brian-brazil commented on src/com/carolinarollergirls/scoreboard/core/Jam.java:
```diff
@@ -0,0 +1,53 @@
+
+    public enum Value implements PermanentProperty {
+        PERIOD_NUMBER(Integer.class, 0),
+        STAR_PASS(Boolean.class, false),
```
This could be named better, it's not obvious how this is modeled.

frank-weinberg replied:
Suggestion?

brian-brazil replied:
Actually a comment that this is if either team had one should do it.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/core/ScoreBoard.java:
```diff
@@ -47,6 +55,14 @@
+    public boolean isInJam();
+    public Jam getUpcomingJam();
+    
+    public void updateTeamJams();
```
A comment would be useful here

brian-brazil commented on src/com/carolinarollergirls/scoreboard/core/ScoreBoard.java:
```diff
@@ -84,26 +96,63 @@
+    public enum Value implements PermanentProperty {
+        CURRENT_PERIOD_NUMBER(Integer.class, 0),
+        CURRENT_PERIOD(Period.class, null),
+        UPCOMING_JAM(Jam.class, null),
```
Why upcoming and not current?

frank-weinberg replied:
Because this will hold a Jam until it becomes current, at which point the jam is moved to the current period and this will hold the next jam.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/event/ScoreBoardEventProviderImpl.java:
```diff
@@ -0,0 +1,446 @@
+        synchronized (coreLock) {
+            @SuppressWarnings("rawtypes")
+            Class type = prop.getType();
+            if (type == TimeoutOwner.class) { return scoreBoard.getTimeoutOwner(sValue); }
```
This is a layering violation, this should come from the object rather than this class

frank-weinberg replied:
I don't understand this comment.

brian-brazil replied:
Check if the fromString method exists via an interface, rather than putting a list of classes here. We should only need to special case the POJOs.

frank-weinberg replied:
Specifically for TimeoutOwner this does not really work. As we don't have an Object yet, the fromString() method needs to be static. But the list of possible `TimeoutOwner`s includes `Team`s, which are only created at runtime in a specific `ScoreBoardImpl` object. So we would need to
a) inject the current `ScoreBoard` instance into a static variable in `TimeoutOwner` so `fromString()` can reference it.
b) use reflection to call the static `fromString()` method (this would also be required for the other classes)
I don't think that is an improvement over what we have now.

brian-brazil replied:
You could pass the scoreBoard object as a parameter to valueFromString. You could also move this all over to the ScoreBoard object.

> I don't think that is an improvement over what we have now.

Two packages mutually depending on each other is to be avoided. The event provider code shouldn't know what types of object it's working with.

frank-weinberg replied:
If we want to hand the `scoreboard` object as a method parameter, it has to come from the `json` resp. `xml` package, two packages that also are referenced from `core` - we would again just shift the circular dependency around (and if we don't remove the other uses of the object in SBEPI, we don't even remove a dependency).

I argue that the `scoreboard` object is at the heart of this program and trying to isolate it in the `core` package is going to cause a huge amount of overhead and will significantly decrease the maintainability of the code, just for the sake of following a principle, as we can be rather certain that this class isn't going away in a refactoring. So this is a case, where a deviation from the principle is justified.

brian-brazil replied:
The only thing even using this method is WS/xml, which can get it from the scoreboard object they already have directly. There's no reason for this code to be here.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/event/ScoreBoardEventProviderImpl.java:
```diff
@@ -0,0 +1,446 @@
+        set(targetProperty, sourceElement.get(sourceProperty), Flag.COPY);
+        return l;
+    }
+    protected ScoreBoardListener setCopy(final PermanentProperty targetProperty, ScoreBoardEventProvider indirectionElement,
```
Comments would be useful to understand what indirection, inverse, copy, recalcuate etc. mean.

brian-brazil replied:
It's still not clear to me what an indirectionProperty is, the comment is mostly repeating the type signature.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/utils/Comparators.java:
```diff
@@ -0,0 +1,71 @@
+
+import java.util.Comparator;
+
+import com.carolinarollergirls.scoreboard.core.BoxTrip;
```
Layering violation, utils shouldn't be using core.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/core/impl/ClockImpl.java:
```diff
@@ -189,62 +193,14 @@ public long getTimeRemaining() {
-            return ms;
-        }
-    }
+    public boolean resetTime() { return set(Value.TIME, getTime(), Flag.RESET); }
```
YYou don't use this return value

brian-brazil reviewed (commented):
This does make things simpler, and I see what you mean about Java 8

brian-brazil committed:
Relax WS message size limit.

This was causing problems such as:

2019-04-01 21:42:02.415:WARN:oejw.WebSocketConnectionRFC6455:Text
message too large > 16384 chars for
SCEP@36de1067{l(/127.0.0.1:55418)<->r(/127.0.0.1:8000),s=1,open=true,ishut=false,oshut=false,rb=false,wb=false,w=true,i=1r}-{WebSocketServletConnectionRFC6455
p=WebSocketParserRFC6455@375f6afb state=DATA buffer=
g=WebSocketGeneratorRFC6455@3e651079 closed=false buffer=-1}

frank-weinberg committed:
Address review comments + minor changes/bugfixes:
- Fix change flag being ignored on +1/-1 buttons (and possibly others)
- If the score is removed in the scoring trip editor, remove the trip
- Ignore attempts to add a box trip to a skater-less Fielding instead of
throwing a NPE

frank-weinberg committed:
remove Java 8 code

frank-weinberg committed:
Merge remote-tracking branch 'upstream/dev' into integrateStats

frank-weinberg commented:
Unittests are in need of a general overhaul anyways. Some of the existing ones have become redundant, some should be in different files now. And most of the new code is not covered at all yet. That's on my list of things to do before a stable 4.x release. (I don't think we need to have them for a beta.)

brian-brazil commented:
For a specific case where there was a bug, a unittest should really be included.

brian-brazil commented on src/com/carolinarollergirls/scoreboard/core/Comparators.java:
```diff
@@ -0,0 +1,63 @@
+    public static Comparator<TeamJam> TeamJamComparator = new SbepComparator<TeamJam>();
+    public static Comparator<ScoringTrip> ScoringTripComparator = new SbepComparator<ScoringTrip>();
+    public static Comparator<Fielding> FieldingComparator = new SbepComparator<Fielding>();
+    public static Comparator<BoxTrip> BoxTripComparator = new Comparator<BoxTrip>() {
```
Digging a bit more, this would better live in the BoxTrip class implementing the Comparable interface: https://docs.oracle.com/javase/7/docs/api/java/lang/Comparable.html

List.sort will use that automatically then.

frank-weinberg committed:
more review comments

frank-weinberg committed:
Merge pull request #1 from brian-brazil/stats-16k-ws

Relax WS message size limit.

frank-weinberg committed:
add PLT screen to the index page

frank-weinberg committed:
add a screen that allows easy copy & paste into a WFTDA Statsbook

frank-weinberg committed:
mark the current scoring trip with a dot instead of a zero when there is
no score entered yet.

frank-weinberg committed:
- add a trip, if points are added during a jam, not only if the score
update fades out during the jam.
- make speed score buttons 1-4 instead of 2-5. closes #216

frank-weinberg committed:
more minor fixes

brian-brazil commented:
:+1: 

I don't 100% agree with everything here and there's known bugs, however the right thing to do is merge and iterate rather than trying to work on a 12k line PR.
